src/CellData/CellFields.jl

```diff
@@ -204,14 +204,97 @@ DomainStyle(::Type{GenericCellField{DS}}) where DS = DS()
 
 (a::CellField)(x) = evaluate(a,x)
 
+struct CellFieldCache
+  kdtree::KDTree{SVector{D,T},Euclidean,T} where {D,T}
+end
+
+function return_cache(f::CellField,x::Point)
+  # Create k-d tree for mesh
+  trian = get_triangulation(f)
+  node_coordinates = get_node_coordinates(trian)
+  @assert !isempty(node_coordinates)
+  sample_coord = node_coordinates[1]
+  D = length(sample_coord)
+  T = eltype(sample_coord)
+  @assert D > 0
+  kdtree = KDTree(map(nc -> SVector(nc.data), node_coordinates))
+  kdtree::KDTree{SVector{D,T},Euclidean,T}
+  return CellFieldCache(kdtree)
+end
+
 function evaluate!(cache,f::CellField,x::Point)
-  @notimplemented """\n
-  Evaluation of a CellField at a given Point is not implemented yet.
+  cache::CellFieldCache
+
+  # Get k-d tree for mesh from cache
+  trian = get_triangulation(f)
+  node_coordinates = get_node_coordinates(trian)
+  @assert !isempty(node_coordinates)
+  sample_coord = node_coordinates[1]
+  D = length(sample_coord)
+  T = eltype(sample_coord)
+  @assert D > 0
+  kdtree = cache.kdtree::KDTree{SVector{D,T},Euclidean,T}
+  
+  # Find neighbouring cells
+  id,dist = nn(kdtree, SVector(x.data))
+  cells = Int[]
+  # TODO: Use better algorithm
+  # Francesc Verdugo @fverdugo 14:07
+  # You have to start with a DiscreteModel, then
+  # topo=get_grid_topology(model)
+  # vertex_to_cells = get_faces(topo,0,num_cell_dims(model))
+  for (c,ids) in enumerate(trian.cell_node_ids)
+    id ∈ ids && push!(cells, c)
+  end
+  @assert !isempty(cells)
+
+  # Find polytope
+  reffes = trian.reffes
+  @assert length(reffes) == 1
+  reffe = reffes[1].reffe
+  polytope = reffe.polytope
+  extrusion = polytope.extrusion
+  @assert length(extrusion) == D
+
+  # Loop over neighbouring cells and calculate distances
+  dists = T[]
+  
+  if all(e == TET_AXIS for e in extrusion)
+    # Calculate barycentric coordinates for a point x in the given cell
+    function bary(cell, x)
+      local ids = trian.cell_node_ids[cell]
+      @assert length(ids) == D+1 "simplex has correct number of vertices"
+      local ys = zero(SVector{D+1,SVector{D,T}})
+      for (n,id) in enumerate(ids)
+        local y = SVector(node_coordinates[id].data)
+        ys = setindex(ys, y, n)
+      end
+      local λ = cartesian2barycentric(ys, SVector(x.data))
+      return λ
+    end
 
-  This is a feature that we want to have at some point in Gridap.
-  If you are ready to help with this implementation, please contact the
-  Gridap administrators.
-  """
+    for c in cells
+      λ = bary(c, x)
+      # Negative distances are outside, positive distance are inside the simplex
+      dist = min(minimum(λ), 1 - maximum(λ))
+      push!(dists, dist)
+    end
+
+    dist, m = findmax(dists)
+    cell = cells[m]
+    λ = bary(cell, x)
+
+    # Ensure the point is inside one of the cells, up to round-off errors
+    @assert dist ≥ -1000 * eps(T) "Point is not inside a cell"
+
```

Comment:
We have to decide what happens if the point is in more than once cell. We can provide just one of the cells. The result is going to be identical for C0 spaces. For discontinuous spaces, the query is ill-posed (the function is not defined at this point) and we would just return the function in one of the cells.

Reply:
I am calculating the distance from the point to all the cells. Without round-off, and with non-overlapping cells, the distance would be positive for exactly one cell and negative for all other ones. (I should probably change my sign convention.) Due to round-off, the distance can be slightly positive or slightly negative for points near cell boundaries, in particular near vertices. In this case, I choose the cell with the largest distance, and I check that the distance (if negative) is at most at round-off level.

I'll update the comments in the code.

Reply:
I think that there is the (unlikely) possibility that the distance will be zero (or in the round-off error) for some cells, am I wrong? I was thinking of cells as closed domains, which is the standard definition in finite elements. If you consider that cells are open domains, you can find points that do not belong to any cell. At the end is just `<` vs `<=`, etc.

Reply:
In any case, what you propose is resilient to this, so keep it as it is

Reply:
I view the cells as closed as well, and would be happy to return a value from any one of the cells if the point lies on the boundary. The problem I'm addressing here is only round-off error, which is a real, practical problem. For example, asking for the value at a vertex can yield distances that say the point is outside of all the neighbouring cells.

For the cases I tested, this error can be more than 100 eps, but was always less than 1000 eps. This distance is in terms of the barycentric coordinates which have a range of 0...1. In this case, the code will extrapolate from one of the cells by that distance, i.e. about 1e-12 of a cell width.